File: sdks/python/mindclade_internal_sdk/datasets.py
"""Private Dataset lifecycle façade over generated protobuf/gRPC contracts."""

from __future__ import annotations

import copy
from dataclasses import replace
from typing import cast

from mindclade.common.v1 import resource_reference_pb2
from mindclade.dataset.v1 import dataset_commands_pb2, dataset_pb2, dataset_release_pb2
from mindclade.internal.dataset.v1 import dataset_service_pb2
from mindclade.operation.v1 import operation_pb2

from ._invocation import AsyncInvoker, SyncInvoker, canonical_digest, command_context
from ._validation import required_response_message, required_text
from .calls import CallOptions, PreparedCall, prepare_call
from .transport import (
    CREATE_DATASET,
    GET_DATASET,
    GET_DATASET_RELEASE,
    LIST_DATASET_RELEASES,
    LIST_DATASETS,
    PUBLISH_DATASET_RELEASE,
    REVOKE_DATASET_RELEASE,
    UPDATE_DATASET,
)


def _project_name(invoker: SyncInvoker | AsyncInvoker) -> str:
    return invoker.config.project_parent


def _project_ref(invoker: SyncInvoker | AsyncInvoker) -> resource_reference_pb2.ResourceRef:
    config = invoker.config
    return resource_reference_pb2.ResourceRef(
        resource_type="project",
        resource_id=config.project_id,
        tenant_id=config.tenant_id,
        project_id=config.project_id,
        name=config.project_parent,
    )


def _mutation_options(command_key: str, options: CallOptions | None) -> CallOptions | None:
    if not command_key or (options is not None and options.idempotency_key is not None):
        return options
    return (
        replace(options, idempotency_key=command_key)
        if options
        else CallOptions(idempotency_key=command_key)
    )


def _dataset_name(invoker: SyncInvoker | AsyncInvoker, name: str) -> str:
    value = required_text("dataset name", name)
    prefix = f"{_project_name(invoker)}/datasets/"
    suffix = value.removeprefix(prefix)
    if not value.startswith(prefix) or not suffix or "/" in suffix:
        raise ValueError("dataset name must be scoped to the configured project")
    return value


def _release_name(invoker: SyncInvoker | AsyncInvoker, name: str) -> str:
    value = required_text("dataset release name", name)
    prefix = f"{_project_name(invoker)}/datasets/"
    suffix = value.removeprefix(prefix)
    parts = suffix.split("/releases/")
    if (
        not value.startswith(prefix)
        or len(parts) != 2
        or not all(parts)
        or any("/" in part for part in parts)
    ):
        raise ValueError("dataset release name must be scoped to the configured project")
    return value


def _normalize_reference(
    invoker: SyncInvoker | AsyncInvoker,
    reference: resource_reference_pb2.ResourceRef,
    *,
    resource_type: str,
    release: bool,
) -> None:
    name = (
        _release_name(invoker, reference.name)
        if release
        else _dataset_name(invoker, reference.name)
    )
    resource_id = name.rsplit("/", 1)[-1]
    config = invoker.config
    if reference.resource_type and reference.resource_type != resource_type:
        raise ValueError("resource reference type does not match the command")
    if reference.resource_id and reference.resource_id != resource_id:
        raise ValueError("resource reference id does not match its name")
    if reference.tenant_id and reference.tenant_id != config.tenant_id:
        raise ValueError("resource reference tenant does not match client scope")
    if reference.project_id and reference.project_id != config.project_id:
        raise ValueError("resource reference project does not match client scope")
    reference.resource_type = resource_type
    reference.resource_id = resource_id
    reference.tenant_id = config.tenant_id
    reference.project_id = config.project_id


def _operation(response: object, *, label: str) -> operation_pb2.Operation:
    value = required_response_message(
        cast(dataset_service_pb2.CreateDatasetResponse, response),
        "operation",
        operation_pb2.Operation,
        label=label,
    )
    required_text("operation id", value.operation_id)
    return value


def _prepare_command[
    CommandT: (
        dataset_commands_pb2.CreateDatasetCommand,
        dataset_commands_pb2.UpdateDatasetCommand,
        dataset_commands_pb2.PublishDatasetReleaseCommand,
        dataset_commands_pb2.RevokeDatasetReleaseCommand,
    )
](
    invoker: SyncInvoker | AsyncInvoker,
    command: CommandT,
    options: CallOptions | None,
) -> tuple[CommandT, PreparedCall]:
    materialized = copy.deepcopy(command)
    key = materialized.context.idempotency_key if materialized.HasField("context") else ""
    materialized.ClearField("context")
    call = prepare_call(
        _mutation_options(key, options),
        default_timeout=invoker.config.default_timeout,
        require_idempotency=True,
    )
    materialized.context.CopyFrom(
        command_context(invoker.config, call, request_digest=canonical_digest(materialized))
    )
    return materialized, call


class Datasets:
    """Synchronous generated-type-only Dataset lifecycle API."""

    def __init__(self, invoker: SyncInvoker) -> None:
        self._invoker = invoker

    def create(
        self,
        command: dataset_commands_pb2.CreateDatasetCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        materialized = dataset_commands_pb2.CreateDatasetCommand()
        materialized.CopyFrom(command)
        if materialized.HasField("project"):
            if materialized.project.name != _project_name(self._invoker):
                raise ValueError("dataset project must match the configured project")
        else:
            materialized.project.CopyFrom(_project_ref(self._invoker))
        materialized, call = _prepare_command(self._invoker, materialized, options)
        response = self._invoker.unary(
            CREATE_DATASET,
            dataset_service_pb2.CreateDatasetRequest(command=materialized),
            call=call,
            retry_safe=True,
        )
        return _operation(response, label="dataset create")

    def get(
        self, name: str, *, if_none_match: str = "", options: CallOptions | None = None
    ) -> dataset_pb2.Dataset:
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            dataset_service_pb2.GetDatasetResponse,
            self._invoker.unary(
                GET_DATASET,
                dataset_service_pb2.GetDatasetRequest(
                    name=_dataset_name(self._invoker, name), if_none_match=if_none_match.strip()
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response, "dataset", dataset_pb2.Dataset, label="dataset get"
        )

    def list(
        self,
        request: dataset_service_pb2.ListDatasetsRequest | None = None,
        *,
        options: CallOptions | None = None,
    ) -> dataset_service_pb2.ListDatasetsResponse:
        materialized = dataset_service_pb2.ListDatasetsRequest()
        if request is not None:
            materialized.CopyFrom(request)
        parent = _project_name(self._invoker)
        if materialized.parent and materialized.parent != parent:
            raise ValueError("dataset list parent must match the configured project")
        materialized.parent = parent
        if materialized.HasField("page") and materialized.page.page_size > 1000:
            raise ValueError("dataset page size cannot exceed 1000")
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        return cast(
            dataset_service_pb2.ListDatasetsResponse,
            self._invoker.unary(LIST_DATASETS, materialized, call=call, retry_safe=True),
        )

    def update(
        self,
        command: dataset_commands_pb2.UpdateDatasetCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if not command.HasField("dataset"):
            raise ValueError("update command requires a generated dataset")
        _dataset_name(self._invoker, command.dataset.name)
        materialized, call = _prepare_command(self._invoker, command, options)
        response = self._invoker.unary(
            UPDATE_DATASET,
            dataset_service_pb2.UpdateDatasetRequest(command=materialized),
            call=call,
            retry_safe=True,
        )
        return _operation(response, label="dataset update")

    def publish_release(
        self,
        command: dataset_commands_pb2.PublishDatasetReleaseCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if not command.HasField("dataset"):
            raise ValueError("publish command requires a generated dataset reference")
        materialized = copy.deepcopy(command)
        _normalize_reference(
            self._invoker, materialized.dataset, resource_type="dataset", release=False
        )
        materialized, call = _prepare_command(self._invoker, materialized, options)
        response = self._invoker.unary(
            PUBLISH_DATASET_RELEASE,
            dataset_service_pb2.PublishDatasetReleaseRequest(command=materialized),
            call=call,
            retry_safe=True,
        )
        return _operation(response, label="dataset release publication")

    def revoke_release(
        self,
        command: dataset_commands_pb2.RevokeDatasetReleaseCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if not command.HasField("dataset_release"):
            raise ValueError("revoke command requires a generated dataset release reference")
        materialized = copy.deepcopy(command)
        _normalize_reference(
            self._invoker,
            materialized.dataset_release,
            resource_type="dataset_release",
            release=True,
        )
        materialized, call = _prepare_command(self._invoker, materialized, options)
        response = self._invoker.unary(
            REVOKE_DATASET_RELEASE,
            dataset_service_pb2.RevokeDatasetReleaseRequest(command=materialized),
            call=call,
            retry_safe=True,
        )
        return _operation(response, label="dataset release revocation")

    def get_release(
        self, name: str, *, options: CallOptions | None = None
    ) -> dataset_release_pb2.DatasetRelease:
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            dataset_service_pb2.GetDatasetReleaseResponse,
            self._invoker.unary(
                GET_DATASET_RELEASE,
                dataset_service_pb2.GetDatasetReleaseRequest(
                    name=_release_name(self._invoker, name)
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response,
            "dataset_release",
            dataset_release_pb2.DatasetRelease,
            label="dataset release get",
        )

    def list_releases(
        self,
        request: dataset_service_pb2.ListDatasetReleasesRequest,
        *,
        options: CallOptions | None = None,
    ) -> dataset_service_pb2.ListDatasetReleasesResponse:
        materialized = dataset_service_pb2.ListDatasetReleasesRequest()
        materialized.CopyFrom(request)
        _dataset_name(self._invoker, materialized.parent)
        if materialized.HasField("page") and materialized.page.page_size > 1000:
            raise ValueError("dataset release page size cannot exceed 1000")
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        return cast(
            dataset_service_pb2.ListDatasetReleasesResponse,
            self._invoker.unary(LIST_DATASET_RELEASES, materialized, call=call, retry_safe=True),
        )


class AsyncDatasets:
    """Asyncio generated-type-only Dataset lifecycle API."""

    def __init__(self, invoker: AsyncInvoker) -> None:
        self._invoker = invoker

    async def create(
        self,
        command: dataset_commands_pb2.CreateDatasetCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        materialized = dataset_commands_pb2.CreateDatasetCommand()
        materialized.CopyFrom(command)
        if materialized.HasField("project") and materialized.project.name != _project_name(
            self._invoker
        ):
            raise ValueError("dataset project must match the configured project")
        if not materialized.HasField("project"):
            materialized.project.CopyFrom(_project_ref(self._invoker))
        materialized, call = _prepare_command(self._invoker, materialized, options)
        return _operation(
            await self._invoker.unary(
                CREATE_DATASET,
                dataset_service_pb2.CreateDatasetRequest(command=materialized),
                call=call,
                retry_safe=True,
            ),
            label="dataset create",
        )

    async def get(
        self, name: str, *, if_none_match: str = "", options: CallOptions | None = None
    ) -> dataset_pb2.Dataset:
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            dataset_service_pb2.GetDatasetResponse,
            await self._invoker.unary(
                GET_DATASET,
                dataset_service_pb2.GetDatasetRequest(
                    name=_dataset_name(self._invoker, name), if_none_match=if_none_match.strip()
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response, "dataset", dataset_pb2.Dataset, label="dataset get"
        )

    async def list(
        self,
        request: dataset_service_pb2.ListDatasetsRequest | None = None,
        *,
        options: CallOptions | None = None,
    ) -> dataset_service_pb2.ListDatasetsResponse:
        materialized = dataset_service_pb2.ListDatasetsRequest()
        materialized.CopyFrom(request or dataset_service_pb2.ListDatasetsRequest())
        parent = _project_name(self._invoker)
        if materialized.parent and materialized.parent != parent:
            raise ValueError("dataset list parent must match the configured project")
        materialized.parent = parent
        if materialized.HasField("page") and materialized.page.page_size > 1000:
            raise ValueError("dataset page size cannot exceed 1000")
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        return cast(
            dataset_service_pb2.ListDatasetsResponse,
            await self._invoker.unary(LIST_DATASETS, materialized, call=call, retry_safe=True),
        )

    async def update(
        self,
        command: dataset_commands_pb2.UpdateDatasetCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if not command.HasField("dataset"):
            raise ValueError("update command requires a generated dataset")
        _dataset_name(self._invoker, command.dataset.name)
        materialized, call = _prepare_command(self._invoker, command, options)
        return _operation(
            await self._invoker.unary(
                UPDATE_DATASET,
                dataset_service_pb2.UpdateDatasetRequest(command=materialized),
                call=call,
                retry_safe=True,
            ),
            label="dataset update",
        )

    async def publish_release(
        self,
        command: dataset_commands_pb2.PublishDatasetReleaseCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if not command.HasField("dataset"):
            raise ValueError("publish command requires a generated dataset reference")
        materialized = copy.deepcopy(command)
        _normalize_reference(
            self._invoker, materialized.dataset, resource_type="dataset", release=False
        )
        materialized, call = _prepare_command(self._invoker, materialized, options)
        return _operation(
            await self._invoker.unary(
                PUBLISH_DATASET_RELEASE,
                dataset_service_pb2.PublishDatasetReleaseRequest(command=materialized),
                call=call,
                retry_safe=True,
            ),
            label="dataset release publication",
        )

    async def revoke_release(
        self,
        command: dataset_commands_pb2.RevokeDatasetReleaseCommand,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if not command.HasField("dataset_release"):
            raise ValueError("revoke command requires a generated dataset release reference")
        materialized = copy.deepcopy(command)
        _normalize_reference(
            self._invoker,
            materialized.dataset_release,
            resource_type="dataset_release",
            release=True,
        )
        materialized, call = _prepare_command(self._invoker, materialized, options)
        return _operation(
            await self._invoker.unary(
                REVOKE_DATASET_RELEASE,
                dataset_service_pb2.RevokeDatasetReleaseRequest(command=materialized),
                call=call,
                retry_safe=True,
            ),
            label="dataset release revocation",
        )

    async def get_release(
        self, name: str, *, options: CallOptions | None = None
    ) -> dataset_release_pb2.DatasetRelease:
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            dataset_service_pb2.GetDatasetReleaseResponse,
            await self._invoker.unary(
                GET_DATASET_RELEASE,
                dataset_service_pb2.GetDatasetReleaseRequest(
                    name=_release_name(self._invoker, name)
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response,
            "dataset_release",
            dataset_release_pb2.DatasetRelease,
            label="dataset release get",
        )

    async def list_releases(
        self,
        request: dataset_service_pb2.ListDatasetReleasesRequest,
        *,
        options: CallOptions | None = None,
    ) -> dataset_service_pb2.ListDatasetReleasesResponse:
        materialized = dataset_service_pb2.ListDatasetReleasesRequest()
        materialized.CopyFrom(request)
        _dataset_name(self._invoker, materialized.parent)
        if materialized.HasField("page") and materialized.page.page_size > 1000:
            raise ValueError("dataset release page size cannot exceed 1000")
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        return cast(
            dataset_service_pb2.ListDatasetReleasesResponse,
            await self._invoker.unary(
                LIST_DATASET_RELEASES, materialized, call=call, retry_safe=True
            ),
        )

File: sdks/go/mindclade/datasets.go
package mindclade

import (
	"context"
	"strings"

	commonv1 "github.com/mindclade/mindclade/protocols/generated/go/common/v1"
	datasetv1 "github.com/mindclade/mindclade/protocols/generated/go/dataset/v1"
	internaldatasetv1 "github.com/mindclade/mindclade/protocols/generated/go/internalrpc/dataset/v1"
	operationv1 "github.com/mindclade/mindclade/protocols/generated/go/operation/v1"
)

// DatasetService is the private lifecycle façade over the generated dataset
// client. Its boundary contains generated protobuf resources and commands
// only; it owns client behavior such as trusted command context, deadlines,
// retries, idempotency, pagination, and normalized errors.
type DatasetService struct {
	client    *Client
	transport internaldatasetv1.DatasetServiceClient
}

func (service *DatasetService) Create(ctx context.Context, command *datasetv1.CreateDatasetCommand, options ...RequestOption) (*operationv1.Operation, error) {
	if command == nil {
		return nil, invalidArgument("a generated CreateDatasetCommand is required")
	}
	materialized := cloneGenerated(command)
	key := materialized.GetContext().GetIdempotencyKey()
	materialized.Context = nil
	if materialized.Project == nil {
		materialized.Project = projectResource(service.client.config)
	} else if materialized.GetProject().GetName() != projectName(service.client.config.TenantID, service.client.config.ProjectID) {
		return nil, invalidArgument("dataset project must match the configured project")
	}
	callContext, request, cancel, err := service.client.mutationContext(ctx, key, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	digest, err := deterministicDigest(materialized)
	if err != nil {
		return nil, err
	}
	materialized.Context = commandContext(service.client.config, callContext, request, digest)
	response, err := service.transport.CreateDataset(callContext, &internaldatasetv1.CreateDatasetRequest{Command: materialized})
	return operationResponse(response.GetOperation(), err, "CreateDataset")
}

func (service *DatasetService) Get(ctx context.Context, name, ifNoneMatch string, options ...RequestOption) (*datasetv1.Dataset, error) {
	if !scopedResourceName(service.client.config, name, "datasets") {
		return nil, invalidArgument("dataset name must be in the configured project")
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.GetDataset(callContext, &internaldatasetv1.GetDatasetRequest{Name: name, IfNoneMatch: strings.TrimSpace(ifNoneMatch)})
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetDataset() == nil {
		return nil, protocolDataLoss("GetDataset response omitted its dataset")
	}
	return cloneGenerated(response.GetDataset()), nil
}

// List returns one generated, opaque-token page. The configured project is
// authoritative; a missing parent is filled and a mismatched parent is rejected.
func (service *DatasetService) List(ctx context.Context, request *internaldatasetv1.ListDatasetsRequest, options ...RequestOption) (*internaldatasetv1.ListDatasetsResponse, error) {
	materialized := cloneGenerated(request)
	if materialized == nil {
		materialized = &internaldatasetv1.ListDatasetsRequest{}
	}
	parent := projectName(service.client.config.TenantID, service.client.config.ProjectID)
	if materialized.GetParent() == "" {
		materialized.Parent = parent
	} else if materialized.GetParent() != parent {
		return nil, invalidArgument("dataset list parent must match the configured project")
	}
	if materialized.GetPage().GetPageSize() > 1000 {
		return nil, invalidArgument("dataset page size cannot exceed 1000")
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.ListDatasets(callContext, materialized)
	if err != nil {
		return nil, normalizeError(err)
	}
	return cloneGenerated(response), nil
}

func (service *DatasetService) Update(ctx context.Context, command *datasetv1.UpdateDatasetCommand, options ...RequestOption) (*operationv1.Operation, error) {
	if command == nil || command.GetDataset() == nil {
		return nil, invalidArgument("a generated UpdateDatasetCommand and dataset are required")
	}
	return service.update(ctx, cloneGenerated(command), options...)
}

func (service *DatasetService) update(ctx context.Context, command *datasetv1.UpdateDatasetCommand, options ...RequestOption) (*operationv1.Operation, error) {
	key := command.GetContext().GetIdempotencyKey()
	command.Context = nil
	if !scopedResourceName(service.client.config, command.GetDataset().GetName(), "datasets") {
		return nil, invalidArgument("updated dataset must be in the configured project")
	}
	callContext, request, cancel, err := service.client.mutationContext(ctx, key, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	digest, err := deterministicDigest(command)
	if err != nil {
		return nil, err
	}
	command.Context = commandContext(service.client.config, callContext, request, digest)
	response, err := service.transport.UpdateDataset(callContext, &internaldatasetv1.UpdateDatasetRequest{Command: command})
	return operationResponse(response.GetOperation(), err, "UpdateDataset")
}

func (service *DatasetService) PublishRelease(ctx context.Context, command *datasetv1.PublishDatasetReleaseCommand, options ...RequestOption) (*operationv1.Operation, error) {
	if command == nil || command.GetDataset() == nil {
		return nil, invalidArgument("a generated PublishDatasetReleaseCommand is required")
	}
	materialized := cloneGenerated(command)
	key := materialized.GetContext().GetIdempotencyKey()
	materialized.Context = nil
	if !normalizeScopedReference(service.client.config, materialized.GetDataset(), "dataset", "datasets") {
		return nil, invalidArgument("dataset release must target the configured project")
	}
	callContext, request, cancel, err := service.client.mutationContext(ctx, key, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	digest, err := deterministicDigest(materialized)
	if err != nil {
		return nil, err
	}
	materialized.Context = commandContext(service.client.config, callContext, request, digest)
	response, err := service.transport.PublishDatasetRelease(callContext, &internaldatasetv1.PublishDatasetReleaseRequest{Command: materialized})
	return operationResponse(response.GetOperation(), err, "PublishDatasetRelease")
}

func (service *DatasetService) RevokeRelease(ctx context.Context, command *datasetv1.RevokeDatasetReleaseCommand, options ...RequestOption) (*operationv1.Operation, error) {
	if command == nil || command.GetDatasetRelease() == nil {
		return nil, invalidArgument("a generated RevokeDatasetReleaseCommand is required")
	}
	materialized := cloneGenerated(command)
	key := materialized.GetContext().GetIdempotencyKey()
	materialized.Context = nil
	if !normalizeScopedReference(service.client.config, materialized.GetDatasetRelease(), "dataset_release", "datasets") {
		return nil, invalidArgument("dataset release must be in the configured project")
	}
	callContext, request, cancel, err := service.client.mutationContext(ctx, key, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	digest, err := deterministicDigest(materialized)
	if err != nil {
		return nil, err
	}
	materialized.Context = commandContext(service.client.config, callContext, request, digest)
	response, err := service.transport.RevokeDatasetRelease(callContext, &internaldatasetv1.RevokeDatasetReleaseRequest{Command: materialized})
	return operationResponse(response.GetOperation(), err, "RevokeDatasetRelease")
}

func (service *DatasetService) GetRelease(ctx context.Context, name string, options ...RequestOption) (*datasetv1.DatasetRelease, error) {
	if !scopedReleaseName(service.client.config, name, "datasets") {
		return nil, invalidArgument("dataset release name must be in the configured project")
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.GetDatasetRelease(callContext, &internaldatasetv1.GetDatasetReleaseRequest{Name: name})
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetDatasetRelease() == nil {
		return nil, protocolDataLoss("GetDatasetRelease response omitted its release")
	}
	return cloneGenerated(response.GetDatasetRelease()), nil
}

func (service *DatasetService) ListReleases(ctx context.Context, request *internaldatasetv1.ListDatasetReleasesRequest, options ...RequestOption) (*internaldatasetv1.ListDatasetReleasesResponse, error) {
	if request == nil || !scopedResourceName(service.client.config, request.GetParent(), "datasets") {
		return nil, invalidArgument("dataset release parent must be a dataset in the configured project")
	}
	materialized := cloneGenerated(request)
	if materialized.GetPage().GetPageSize() > 1000 {
		return nil, invalidArgument("dataset release page size cannot exceed 1000")
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.ListDatasetReleases(callContext, materialized)
	if err != nil {
		return nil, normalizeError(err)
	}
	return cloneGenerated(response), nil
}

func projectResource(config Config) *commonv1.ResourceRef {
	name := projectName(config.TenantID, config.ProjectID)
	return &commonv1.ResourceRef{ResourceType: "project", ResourceId: config.ProjectID, TenantId: config.TenantID, ProjectId: config.ProjectID, Name: name}
}

func operationResponse(operation *operationv1.Operation, err error, method string) (*operationv1.Operation, error) {
	if err != nil {
		return nil, normalizeError(err)
	}
	if operation == nil || strings.TrimSpace(operation.GetOperationId()) == "" {
		return nil, protocolDataLoss(method + " response omitted its durable operation")
	}
	return cloneGenerated(operation), nil
}

func scopedResourceName(config Config, name, collection string) bool {
	prefix := projectName(config.TenantID, config.ProjectID) + "/" + collection + "/"
	remainder := strings.TrimPrefix(strings.TrimSpace(name), prefix)
	return strings.HasPrefix(strings.TrimSpace(name), prefix) && remainder != "" && !strings.Contains(remainder, "/") && validResourceIdentifier(remainder)
}

func scopedReleaseName(config Config, name, collection string) bool {
	prefix := projectName(config.TenantID, config.ProjectID) + "/" + collection + "/"
	remainder := strings.TrimPrefix(strings.TrimSpace(name), prefix)
	parts := strings.Split(remainder, "/releases/")
	return strings.HasPrefix(strings.TrimSpace(name), prefix) && len(parts) == 2 && validResourceIdentifier(parts[0]) && validResourceIdentifier(parts[1]) && !strings.Contains(parts[0], "/") && !strings.Contains(parts[1], "/")
}

func normalizeScopedReference(config Config, reference *commonv1.ResourceRef, resourceType, collection string) bool {
	if reference == nil || (reference.GetResourceType() != "" && reference.GetResourceType() != resourceType) ||
		(reference.GetTenantId() != "" && reference.GetTenantId() != config.TenantID) ||
		(reference.GetProjectId() != "" && reference.GetProjectId() != config.ProjectID) {
		return false
	}
	validName := scopedResourceName(config, reference.GetName(), collection)
	if strings.HasSuffix(resourceType, "_release") {
		validName = scopedReleaseName(config, reference.GetName(), collection)
	}
	if !validName {
		return false
	}
	parts := strings.Split(reference.GetName(), "/")
	resourceID := parts[len(parts)-1]
	if reference.GetResourceId() != "" && reference.GetResourceId() != resourceID {
		return false
	}
	reference.ResourceType = resourceType
	reference.ResourceId = resourceID
	reference.TenantId = config.TenantID
	reference.ProjectId = config.ProjectID
	return true
}

func invalidArgument(message string) error {
	return &Error{Code: CodeInvalidArgument, Message: message}
}

func protocolDataLoss(message string) error {
	return &Error{Code: CodeDataLoss, Message: message}
}
